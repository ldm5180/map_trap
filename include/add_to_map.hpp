#ifndef ADD_TO_MAP_HPP
#define ADD_TO_MAP_HPP

#include <algorithm>
#include <benchmark/benchmark.h>
#include <cstdlib>
#include <string>

static const auto add_to_map = [](const auto num_elems, auto &map) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
  }
  benchmark::ClobberMemory();
};

static const auto add_to_map_and_key_list = [](const auto num_elems, auto &map,
                                               auto &keys) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
    keys.push_back(random_key);
    std::sort(keys.begin(), keys.end());
  }
  benchmark::ClobberMemory();
};

#endif // ADD_TO_MAP_HPP
