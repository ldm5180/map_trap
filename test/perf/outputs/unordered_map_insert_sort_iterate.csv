Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:42:53
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_insert_sort_iterate/8",495443,1358.88,1358.75,ns,,,,,
"BM_unordered_map_insert_sort_iterate/16",119376,6206.18,6205.97,ns,,,,,
"BM_unordered_map_insert_sort_iterate/32",26146,26887.7,26884.7,ns,,,,,
"BM_unordered_map_insert_sort_iterate/64",5056,133672,133667,ns,,,,,
"BM_unordered_map_insert_sort_iterate/128",1185,616580,616480,ns,,,,,
