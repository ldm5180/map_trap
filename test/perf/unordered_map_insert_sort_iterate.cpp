#include "add_to_map.hpp"

#include <benchmark/benchmark.h>
#include <cstdlib>
#include <unordered_map>

using map_t = std::unordered_map<int64_t, int64_t>;

static void BM_unordered_map_insert_sort_iterate(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  map.reserve(state.range(0));
  std::vector<map_t::key_type> keys{};
  keys.reserve(state.range(0));

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map_and_key_list(state.range(0), map, keys);
  }

  for (auto _ : state) {
    map_t::mapped_type v{};
    for (const auto &k : keys) {
      benchmark::DoNotOptimize(v = map[k]);
    }
  }
}
BENCHMARK(BM_unordered_map_insert_sort_iterate)
    ->RangeMultiplier(2)
    ->Range(8, 8 << 4);
BENCHMARK_MAIN();
