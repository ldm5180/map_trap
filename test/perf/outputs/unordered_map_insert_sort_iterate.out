Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:31:15
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_insert_sort_iterate/8",481702,1447.66,1446.42,ns,,,,,
"BM_unordered_map_insert_sort_iterate/16",107894,6146.29,6147.08,ns,,,,,
"BM_unordered_map_insert_sort_iterate/32",24604,26582,26583.3,ns,,,,,
"BM_unordered_map_insert_sort_iterate/64",4782,133809,133860,ns,,,,,
"BM_unordered_map_insert_sort_iterate/128",1108,626719,627033,ns,,,,,
